java-frontend/src/main/java/org/sonar/java/se/checks/InvariantReturnCheck.java

```diff
@@ -0,0 +1,174 @@
+      return methodTree.is(Tree.Kind.CONSTRUCTOR) || returnType.symbolType().isVoid();
+    }
+
+    private static List<ReturnStatementTree> extractReturnStatements(MethodTree methodTree) {
+      ReturnExtractor visitor = new ReturnExtractor();
+      methodTree.accept(visitor);
+      return visitor.returns;
+    }
+  }
+
+  private Deque<MethodInvariantContext> methodInvariantContexts = new LinkedList<>();
+
+  @Override
+  public void init(MethodTree methodTree, CFG cfg) {
+    methodInvariantContexts.push(new MethodInvariantContext(methodTree));
+  }
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    for (SEIssue seIssue : issues) {
+      context.reportIssueWithFlow(this, seIssue.getTree(), seIssue.getMessage(), seIssue.getFlows(), seIssue.getFlows().iterator().next().size());
+    }
+    issues.clear();
+  }
+
+  @Override
+  public void checkEndOfExecutionPath(CheckerContext context, ConstraintManager constraintManager) {
+    MethodInvariantContext methodInvariantContext = methodInvariantContexts.peek();
+    if (!methodInvariantContext.methodToCheck) {
+      return;
+    }
+    SymbolicValue exitValue = context.getState().exitValue();
+
+    if (exitValue != null) {
+      methodInvariantContext.endPaths++;
+      methodInvariantContext.symbolicValues.add(exitValue);
+      PMap<Class<? extends Constraint>, Constraint> constraints = context.getState().getConstraints(exitValue);
+      if (constraints != null) {
+        constraints.forEach(methodInvariantContext.methodConstraints::put);
+      } else {
+        // Relational SV or NOT SV : we can't say anything.
+        methodInvariantContext.avoidRaisingConstraintIssue = true;
+      }
+    }
+  }
+
+  @Override
+  public void checkEndOfExecution(CheckerContext context) {
+    reportIssues();
+  }
+
+  @Override
+  public void interruptedExecution(CheckerContext context) {
+    // pop but do nothing with it : can't report on incomplete execution
+    methodInvariantContexts.pop();
+  }
+
+  private void reportIssues() {
+    MethodInvariantContext methodInvariantContext = methodInvariantContexts.pop();
+    if (!methodInvariantContext.methodToCheck) {
+      return;
+    }
+    if (methodInvariantContext.returnImmutableType && methodInvariantContext.symbolicValues.size() == 1 && methodInvariantContext.endPaths > 1) {
+      report(methodInvariantContext);
+    } else if (!methodInvariantContext.avoidRaisingConstraintIssue) {
+      for (Class<? extends Constraint> constraintClass : methodInvariantContext.methodConstraints.keys()) {
+        Collection<Constraint> constraints = methodInvariantContext.methodConstraints.get(constraintClass);
+        Constraint firstConstraint = constraints.iterator().next();
+        if(constraints.size() == methodInvariantContext.endPaths
+          && constraints.stream().allMatch(c ->firstConstraint.hasPreciseValue() && firstConstraint.equals(c))){
```

Comment:
In fact I think this whole test can be simplified. By using a Set-based multimap (instead of a list) for the constraints, you may be able to raise issue by simply checking if the collection has **size one** and if the **single item has precise value**.